src/main/java/com/hubspot/jinjava/util/EagerReconstructionUtils.java

```diff
@@ -190,89 +121,214 @@ public static EagerExecutionResult executeInChildContext(
+    // This creates a stringified snapshot of the context
+    // so it can be disabled via the config because it may cause performance issues.
+    Stream<Entry<String, Object>> entryStream;
+    if (!interpreter.getConfig().getExecutionMode().useEagerContextReverting()) {
+      entryStream =
         interpreter
           .getContext()
+          .getCombinedScope()
           .entrySet()
           .stream()
-          .filter(e -> initiallyResolvedHashes.containsKey(e.getKey()))
+          .filter(entry -> initiallyResolvedHashes.containsKey(entry.getKey()))
           .filter(
-            e ->
-              !initiallyResolvedHashes
-                .get(e.getKey())
-                .equals(getObjectOrHashCode(e.getValue()))
-          )
-          .collect(
-            Collectors.toMap(
-              Entry::getKey,
-              e -> {
-                if (eagerChildContextConfig.takeNewValue) {
-                  if (e.getValue() instanceof DeferredValue) {
-                    return ((DeferredValue) e.getValue()).getOriginalValue();
-                  }
-                  return e.getValue();
-                }
-
-                if (
-                  e.getValue() instanceof DeferredValue &&
-                  initiallyResolvedHashes
-                    .get(e.getKey())
-                    .equals(
-                      getObjectOrHashCode(
-                        ((DeferredValue) e.getValue()).getOriginalValue()
-                      )
-                    )
-                ) {
-                  return ((DeferredValue) e.getValue()).getOriginalValue();
-                }
-
-                // This is necessary if a state-changing function, such as .update()
-                // or .append() is run against a variable in the context.
-                // It will revert the effects when takeNewValue is false.
-                if (initiallyResolvedAsStrings.containsKey(e.getKey())) {
-                  // convert to new list or map
-                  try {
-                    return interpreter.resolveELExpression(
-                      initiallyResolvedAsStrings.get(e.getKey()),
-                      interpreter.getLineNumber()
-                    );
-                  } catch (DeferredValueException ignored) {}
-                }
-
-                // Previous value could not be mapped to a string
-                throw new DeferredValueException(e.getKey());
-              }
-            )
-          )
-      );
+            entry -> EagerExpressionResolver.isResolvableObject(entry.getValue(), 4, 400) // TODO make this configurable
+          );
+    } else {
+      entryStream =
+        interpreter
+          .getContext()
+          .entrySet()
+          .stream()
+          .filter(entry -> initiallyResolvedHashes.containsKey(entry.getKey()))
+          .filter(
+            entry -> EagerExpressionResolver.isResolvableObject(entry.getValue(), 4, 400) // TODO make this configurable
+          );
```

Comment:
Using something like:
```
initiallyResolvedHashes.keySet()
          .stream()
          .map(key -> new AbstractMap.SimpleImmutableEntry(key, interpreter.getContext().get(key))
          .filter(
            entry -> EagerExpressionResolver.isResolvableObject(entry.getValue(), 4, 400) // TODO make this configurable
          );
```
May be faster too.
Or, simply storing `interpreter.getContext().entrySet()` as a variable in the beginning of this method so that we only need to go through the scope map once in the whole method.

Reply:
> we only need to go through the scope map once in the whole method

Do you mean the `executeInChildContext` method? I think the `before` and `after` the `function.apply`, the entrySet of the context may have changed, we need to go through the scopes again.

Reply:
https://github.com/HubSpot/jinjava/pull/974/commits/9a912af026e807a3019b298b1700a9f78f97697d is exactly what I meant 👍